modules/asset-manager-static-file-authorization/src/main/java/org/opencastproject/assetmanager/auth/AssetManagerStaticFileAuthorization.java

```diff
@@ -0,0 +1,182 @@
+
+  protected EntityManagerFactory entityManagerFactory;
+  private SecurityService securityService;
+
+  private Pattern staticFilePattern = Pattern.compile("^/([^/]+)/(?:api|internal)/([^/]+)/.*$");
+
+  // Settings for role filter
+  private boolean includeAPIRoles = false;
+  private boolean includeCARoles = false;
+  private boolean includeUIRoles = false;
+
+  @Reference
+  public void setEntityManagerFactory(EntityManagerFactory entityManagerFactory) {
+    this.entityManagerFactory = entityManagerFactory;
+  }
+
+  @Reference
+  void setSecurityService(SecurityService securityService) {
+    this.securityService = securityService;
+  }
+
+  @Activate
+  public void activate(ComponentContext cc) {
+    List<Pattern> newPattern = new ArrayList<>();
+    Dictionary<String, Object> properties = cc != null ? cc.getProperties() : new Hashtable<>();
+    staticFilePattern = Pattern.compile(Objects.toString(
+        properties.get("pattern"),
+        "^/([^/]+)/(?:api|internal)/([^/]+)/.*$"));
+    includeAPIRoles = BooleanUtils.toBoolean(Objects.toString(properties.get("evaluate.roles.api"), null));
+    includeCARoles = BooleanUtils.toBoolean(Objects.toString(properties.get("evaluate.roles.ca"), null));
+    includeUIRoles = BooleanUtils.toBoolean(Objects.toString(properties.get("evaluate.roles.ui"), null));
+    logger.info("Started authentication handler for {}", staticFilePattern);
+  }
+
+  @Override
+  public List<Pattern> getProtectedUrlPattern() {
+    return Collections.singletonList(staticFilePattern);
+  }
+
+  @Override
+  public boolean verifyUrlAccess(final String path) {
+    // Always allow access for admin
+    final User user = securityService.getUser();
+    if (user.hasRole(GLOBAL_ADMIN_ROLE)) {
+      logger.debug("Allow access for admin `{}`", user);
+      return true;
+    }
+
+    // Check pattern
+    final Matcher m = staticFilePattern.matcher(path);
+    if (!m.matches()) {
+      logger.debug("Path does not match pattern. Preventing access.");
+      return false;
+    }
+
+    // Check organization
+    final String organizationId = m.group(1);
+    if (!securityService.getOrganization().getId().equals(organizationId)) {
+      logger.debug("The user's organization does not match. Preventing access.");
+      return false;
+    }
+
+    if (user.getRoles().size() == 0) {
+      logger.debug("User has no roles allowing access.");
+      return false;
+    }
+
+    // Check role access
+    final List<String> roles = user.getRoles().parallelStream()
+        .map(Role::getName)
+        .filter(roleFilter)
+        .map((role) -> role + " | read")
+        .collect(Collectors.toList());  // ["ROLE_XY | read", ...]
+
+    StringBuilder properties = new StringBuilder("property_name = ?");
+    for (int i = 1; i < roles.size(); i++) {
+      properties.append(" or property_name = ?");
+    }
+    String sql = "select count(1) from oc_assets_properties "
```

Comment:
Basically two reasons:

1. This cannot depend directly on the asset manager since this runs on different nodes and using a remote for this would have made it way too slow. So this would have basically meant re-implementing the whole thing for a single query or splitting the whole asset manager apart to get a usable base.
2. This is so much faster than what the asset manager does. Speed doesn't really matter there, but here it does since we do real-time requests for checking access to a lot of assets instead of the occasional asset during workflows.
3. This is simple and easy to understand once you get over the somewhat weird `ROLE_XY | read` entries we cannot change since they are already baked into the asset manager.

Reply:
Ok.  I'm going to add a note to this code, and the asset properties object, linking to this discussion.  FTR, the reason I'm leery of this is that it opens the door to other developers doing things like this - especially with queries which write to the DB.  We had quite the discussion way back in the pre-1.0 days with some developers pushing hard for hand crafted SQL, and other pushing back for stability and portability.

In this one case I can see a good, solid reason to do this, but without a similarly compelling reason (ie, needs a massive reimpl, plus perf considerations) I would be *strongly* against further raw queries like this.

Reply:
I agree. I guess, what we mostly need to prevent against are module boundary violations. For example, a statistics service accessing the asset manager table to get data. That could easily become a mess when you change asset manager internals in the future. Writing would be even worst since you are then tinkering with the internal state of another service which could have very weird side effects.

I see this as something slightly different since it is still marked as an asset manager module. As such, accessing the internal state of the asset manager should not be completely unexpected and given the way this works, it should be safe.

But putting that into writing is probably a good idea. I will add a proper explanation as comment to this.

Reply:
Added an explanation:

https://github.com/opencast/opencast/blob/d4086846f3cd8f27f3f5691157bf88fad7c7a884/modules/asset-manager-static-file-authorization/src/main/java/org/opencastproject/assetmanager/auth/AssetManagerStaticFileAuthorization.java#L135-L145